### DISCUSSION TITLE
Stepper motors at 1ms with ESP32VROOM32E

### DISCUSSION THREAD

**Discussion by JeremyLaurenson** [HomeSpan/HomeSpan#782] [Q&A]
I need to run a stepper at 1ms between pulses and have rewritten my own driver to drive the 6600 stepper driver.
The driver works ok, but with homespun does not seem to be able to keep up with the 1ms requirement I have.
It gets 'jerkey' and misses some steps sometimes.

```
stepper1=new Stepper_TB6600(s1in1,s1in2,-1,{2,0});
stepper1->setAccel(5,20);       

```

I am prioritizing the thread, but I am unclear on how to squeeze more performance out of this....

Suggestions?

```

#pragma once

// Implementation of StepperControl for a TB6600 stepper driver
// https://www.dfrobot.com/product-1547.html

// This implementation uses direction and pulse pins. No enable pins only to save pinouts

//////////////////////////
 
struct Stepper_TB6600 : StepperControl {

  int stepPin;
  int dirPin;
  int enablePin;

//////////////////////////

  Stepper_TB6600(int stepPin, int dirPin, int enablePin, std::pair<uint32_t, uint32_t> taskParams = {1,0}) : StepperControl(taskParams.first,taskParams.second){
    this->stepPin=stepPin;
    this->dirPin=dirPin;
    this->enablePin=enablePin;

    pinMode(stepPin,OUTPUT);
    pinMode(dirPin,OUTPUT);
    if(enablePin>0)pinMode(enablePin,OUTPUT);

    setStepType(EIGHTH_STEP);
    //   FULL_STEP_ONE_PHASE=0,
    //FULL_STEP_TWO_PHASE=1,
    //HALF_STEP=2,
    //QUARTER_STEP=4,
    //EIGHTH_STEP=8
  }

//////////////////////////

  void onStep(boolean direction) override {
    digitalWrite(dirPin,direction);
    digitalWrite(stepPin,HIGH);
    delayMicroseconds(20); // Wait for controller to see this.
    digitalWrite(stepPin,LOW);      
  }

//////////////////////////

  void onEnable() override {
    if(enablePin>0)digitalWrite(enablePin,0);
  }

//////////////////////////

  void onDisable() override {
    if(enablePin>0)digitalWrite(enablePin,1);
  }

//////////////////////////

  StepperControl *setStepType(int mode) override {
    switch(mode){
      case FULL_STEP_TWO_PHASE:
        break;
      default:
        ESP_LOGE(STEPPER_TAG,"Unknown StepType=%d",mode);
    }
    return(this);
  }
  
};

//////////////////////////
```

**Comment by HomeSpan** [comment]
You should not need the 20mic delay. I think simply toggling high/low without any delay should suffice (well, at least worth a try).   

As for 1ms, have you separately confirmed the motor can smoothly operate at that speed (independently of HomeSpan)?

**Reply by JeremyLaurenson** [reply]
20mic is needed or the controller doesnt see the state change.
Yes, it runs smoothly in a simple sketch

**Comment by JeremyLaurenson** [comment]
Im thinking of gearing the motor at 2:1 and slowing this all down....

**Reply by HomeSpan** [reply]
I'll see if there is something in my code that is creating any bottlenecks.  May be a result of the way I create the time delay (using millis instead of micros).

**Reply by JeremyLaurenson** [reply]
Hows this looking, Gregg?

**Reply by HomeSpan** [reply]
I've gone back through the logic in my code, and based on the way I structured Stepper as an independent RTOS task, I don't think anything less than 2ms delay will work.  This is because I rely on RTOS to handle task switching so that it optimally shares resources between Stepper, HomeSpan, WiFi, etc.  The only thing my task needs to do is make sure it yields periodically to allow the operating system to switch tasks.  This is the primary purpose of the vTaskDelay() function.  However, this method limits the resolution of task switching to 1 RTOS tick, which is set to 1ms in the Arduino-ESP32 library (it's actually 10ms in the ESP-IDF, which really limits the frequency of task switching).   Since the lowest resolution is 1ms, setting vTaskDelay() to 1ms causes jitter as sometimes the task is skipped altogether.

There is another version of vTaskDelay() that I will explore to see if it will work with a 1ms delay, but I'm doubtful.

I may explore other options in the future, such as a specialized version of Stepper for use in dual-processor chips (the ESP32 and ESP32-S3) where the task does not yield at all (making it possible to add a blocking delay of 1000 microseconds or lower).  However, this will be a back-burner project for now.

**Comment by JeremyLaurenson** [comment]
For those interested, I use an ATTiny1616 controller and had it talk to to Homespun in order to offload this function. The ESP32 sends I2C commands tot he Attiny, which uses a single digital out to connect to the homespun board and signal its completed the action.

**Reply by HomeSpan** [reply]
Thanks for the update.  Using a dedicated driver is probably the most robust solution and makes a lot of sense.
